components/calendar/src/types.rs

```diff
@@ -7,11 +7,42 @@
 #[doc(no_inline)]
 pub use calendrical_calculations::rata_die::RataDie;
 use core::fmt;
+use core::num::NonZeroU8;
 use tinystr::TinyAsciiStr;
 use tinystr::{TinyStr16, TinyStr4};
 use zerovec::maps::ZeroMapKV;
 use zerovec::ule::AsULE;
 
+/// A bag of various ways of expressing the year, month, and/or day.
+///
+/// Pass this into [`Date::try_from_fields`](crate::Date::try_from_fields).
+#[derive(Copy, Clone, Debug, PartialEq, Default)]
+#[non_exhaustive]
+pub struct DateFields<'a> {
+    /// The era code as defined by CLDR.
+    ///
+    /// If set, [`Self::era_year`] must also be set.
+    pub era: Option<&'a str>,
+    /// The numeric year in [`Self::era`].
+    ///
+    /// If set, [`Self::era`] must also be set.
+    pub era_year: Option<i32>,
+    /// See [`Date::extended_year()`](crate::Date::extended_year).
+    ///
+    /// If both this and [`Self::era`]/[`Self::era_year`] are set, they must
+    /// refer to the same year.
+    pub extended_year: Option<i32>,
+    /// The [`MonthCode`] representing a valid month in this calendar year.
+    pub month_code: Option<MonthCode>,
+    /// See [`MonthInfo::ordinal`].
+    ///
+    /// If both this and [`Self::month_code`] are set, they must refer to
+    /// the same month.
+    pub ordinal_month: Option<NonZeroU8>,
```

Comment:
issue: we've repeatedly decided against using `NonZero` on our 1-based APIs, because of the friction on the caller side. you have to handle the values 13..255 anyway, so excluding 0 doesn't gain you much

Reply:
I'm fine with `Option<u8>`

Reply:
I used NonZero here because in Temporal the nonzeroness is validated before reaching this function, but technically it doesn't add anything to rust users

Reply:
Yeah let's not design our API around temporal implementation details

Reply:
It also serves to emphasize that the month and day are 1-based

Reply:
#7009